lens/private/list/flatten.rkt

```diff
@@ -0,0 +1,204 @@
+#lang racket/base
+
+(provide append*-lens flatten/depth-lens flatten/depth unflatten/depth)
+
+(require fancy-app lens racket/list racket/match)
+
+(module+ test
+  (require rackunit syntax/parse lens/private/test-util/test-lens))
+
+;; restore-structure : (Listof (Listof A)) (Listof B) -> (Listof (Listof B))
+;; Takes a list of lists and a list and un-flattens the flattened
+;; argument according to the structure of the structure arguement.
+;; The length of the flattened list must be the same as the length
+;; of (append* structure).
+(define (restore-structure structure flattened)
+  (match structure
+    [(list)
+     (unless (empty? flattened)
+       (error 'append*-lens "flattened list is too long to match the structure"))
```

Comment:
A better idea might be to make another function that calls this one, but first checks that the length of `flattened` is equal to the length of `(append* structure)`. Then I could drop this check.